lib/datura/elasticsearch/index.rb

```diff
@@ -114,19 +124,43 @@ def set_schema
 
   # doc: ruby hash corresponding with Elasticsearch document JSON
   def valid_document?(doc)
-    get_schema_mapping if @schema_mapping.nil?
-    fields = @schema_mapping["fields"]
-    dynamic = @schema_mapping["dynamic"]
+    get_schema_mapping if !defined?(@schema_mapping)
     # NOTE: validation only checking the names of fields
     # against the schema, NOT the contents of fields
-    # since Elasticsearch itself will know if you are sending it
-    # text instead of a date field, etc
-
-    valid = doc.keys.all? do |doc_field|
-      # check if exact match for a field or if it matches dynamic field mapping
-      fields.include?(doc_field) || doc_field.match(dynamic)
+    # Elasticsearch itself checks that you are sending date
+    # formats to date fields, etc
+
+    doc.all? do |field, value|
+      if valid_field?(field)
+        # great, the field is valid, now check if it is a parent
+        nested = Array(value).map do |nested|
+          if nested.class == Hash
+            nested.keys.all? { |k| valid_field?(k, field) }
+          end
+        end
+        # if the array is empty, ignore it, otherwise find out if any
+        # nested fields failed the validate
+        nested.compact.all? { |t| t }
```

Comment:
Couldn't you just return early above as soon as a nested field had an invalid field? It doesn't look like you're returning an array if more than one field is invalid by letting it enumerate all the keys. This all still just returns a single `false`. This could save time as I'm guessing that validating each document adds a little bit more processing time during posting.

Otherwise we could return the field names if they are invalid so we know where to look in the document when it is rejected? That might be extra complicated for programming for little benefit though if we can just spot check the document and find what's goofy.

Reply:
I did Greg's first suggestion